timelock-agent/src/main/java/com/palantir/atlasdb/timelock/paxos/LeadershipComponents.java

```diff
@@ -0,0 +1,160 @@
+    }
+
+    public void shutdown() {
+        closer.shutdown();
+    }
+
+    public LeaderPingHealthCheck healthCheck() {
+        return leaderPingHealthCheck;
+    }
+
+    private LeadershipContext getOrCreateNewLeadershipContext(Client client) {
+        return leadershipContextByClient.computeIfAbsent(client, this::createNewLeadershipContext);
+    }
+
+    private LeadershipContext createNewLeadershipContext(Client client) {
+        LeadershipContext uninstrumentedLeadershipContext = leadershipContextFactory.create(client);
+        closer.register(uninstrumentedLeadershipContext.closeables());
+
+        BatchingLeaderElectionService batchingLeaderElectionService =
+                new BatchingLeaderElectionService(uninstrumentedLeadershipContext.leaderElectionService());
+        closer.register(batchingLeaderElectionService);
+
+        LeaderElectionService leaderElectionService = metrics.instrument(
+                LeaderElectionService.class,
+                batchingLeaderElectionService,
+                "leader-election-service",
+                client);
+        closer.register(() -> shutdownLeaderElectionService(leaderElectionService));
+
+        return ImmutableLeadershipContext.builder()
+                .from(uninstrumentedLeadershipContext)
+                .leaderElectionService(leaderElectionService)
+                .build();
+    }
+
+    private static void shutdownLeaderElectionService(LeaderElectionService leaderElectionService) {
+        leaderElectionService.markNotEligibleForLeadership();
+        leaderElectionService.stepDown();
+    }
+
+    private static class ShutdownAwareCloser {
+        private boolean isShutdown = false;
+        private final Closer closer = Closer.create();
+
+        synchronized void register(Closeable closeable) {
+            register(ImmutableList.of(closeable));
+        }
+
+        /**
+         * Attempts to register a collection of {@link Closeable}s to be closed when shutting down Timelock.
+         * If timelock has already been shutdown, any {@link Closeable}s passed here will be <em>immediately</em>
+         * closed, and a {@link NotCurrentLeaderException} will be thrown.
+         *
+         * @param closeables collection of {@link Closeable}s to be closed when timelock shuts down.
+         */
+        synchronized void register(Collection<Closeable> closeables) {
+            if (isShutdown) {
+                ShutdownAwareCloser immediateCloser = new ShutdownAwareCloser();
+                immediateCloser.register(closeables);
+                immediateCloser.shutdown();
+                throw new NotCurrentLeaderException("This timelock node is being shutdown");
+            } else {
+                closeables.forEach(closer::register);
+            }
+        }
+
+        /**
+         * This is to be called when timelock is shutting down. It will close in LIFO order any resources that were
+         * registered during their creation.
```

Comment:
sure, but naturally you want to close resources in that manner, closing a resources dependant resources makes things fail in spectacular ways + would rather not have to defensively deal with that in each resource

Reply:
Yep, agreed - I'd be ok with relying on that in the code, just that saying it will do that when the closer doesn't actually in its API guarantee that feels a bit much. I think I just wasn't comfortable with saying this explicitly in the docs